### PR TITLE
Add support for reading from Unity catalog with Iceberg REST

### LINKED ISSUE #22609: Issue: Integration with Trino and Unity Catalog
Start Unity Catalog (UC) server:
```
bin/start-uc-server

```
bin/uc table list --catalog unity --schema default
```
┌────────────┬────────────────┬────────────────┬────────────────┬────────────────┬────────────────┬────────────────┬────────────────┬────────────────┬────────────────┬────────────────┬────────────────┐
│    NAME    │  CATALOG_NAME  │  SCHEMA_NAME   │   TABLE_TYPE   │DATA_SOURCE_FORM│    COLUMNS     │STORAGE_LOCATION│    COMMENT     │   PROPERTIES   │   CREATED_AT   │   UPDATED_AT   │    TABLE_ID    │
├────────────┼────────────────┼────────────────┼────────────────┼────────────────┼────────────────┼────────────────┼────────────────┼────────────────┼────────────────┼────────────────┼────────────────┤
│hudi_table  │unity           │default         │EXTERNAL        │DELTA           │[]              │s3://soumilsh...│null            │{}              │1720212320756   │1720212320756   │2b381c29-01fb...│
├────────────┼────────────────┼────────────────┼────────────────┼────────────────┼────────────────┼────────────────┼────────────────┼────────────────┼────────────────┼────────────────┼────────────────┤
│numbers     │unity           │default         │EXTERNAL        │DELTA           │[]              │file:///Users...│null            │{}              │1720097124777   │1720097124777   │6dd0814f-5831...│
└────────────┴────────────────┴────────────────┴────────────────┴────────────────┴────────────────┴────────────────┴────────────────┴────────────────┴────────────────┴────────────────┴────────────────┘

```

Configure Docker Compose:
```
version: "3"
services:
  trino-coordinator:
    image: 'trinodb/trino:latest'
    hostname: trino-coordinator
    ports:
      - '9000:9000'
    volumes:
      - ./trino/etc:/etc/trino

```

Screenshot 2024-07-07 at 9 29 39 AM

Setup SSH tunneling:
```
ssh -R 8080:localhost:8080 abc.serveo.net

```

iceberg.properties
```
connector.name=iceberg
iceberg.catalog.type=rest
iceberg.rest-catalog.uri=http://abc.serveo.net:8080/api/2.1/unity-catalog/iceberg
iceberg.rest-catalog.security=OAUTH2
iceberg.rest-catalog.oauth2.token=not_used
```

Execute into Trino container and connect:
```
trino --server localhost:9000 --catalog iceberg --schema default

```
image

#### SHOW TABLES FROM unity;
```
Query 20240707_132801_00006_qqd25, FAILED, 1 node
Splits: 19 total, 0 done (0.00%)
0.52 [0 rows, 0B] [0 rows/s, 0B/s]

Query 20240707_132801_00006_qqd25 failed: Error listing tables for catalog iceberg: Failed to list tables from namespace: unity

```

Expected Behavior

The tables registered in Unity Catalog should be listed successfully in Trino when querying SHOW TABLES FROM unity;.

Additional Information

Unity Catalog is running at http://abc.serveo.net:8080/.
Followed the documentation provided at: [Unity Catalog Trino Integration Guide](https://github.com/unitycatalog/unitycatalog/blob/main/docs/integrations/unity-catalog-trino.md).

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

This PR supports only read operations because Iceberg REST interface in Unity catalog doesn't support creating tables yet. 

Note that Databricks Unity catalog provides HTTP interface, but it doesn’t return Iceberg UniForm tables. 

Example catalog config properties:
```properties
connector.name=iceberg
iceberg.catalog.type=rest
iceberg.rest-catalog.uri=https://dbc-12345678-9999.cloud.databricks.com/api/2.1/unity-catalog/iceberg
iceberg.security=read_only
iceberg.rest-catalog.security=OAUTH2
iceberg.rest-catalog.oauth2.token=***
iceberg.rest-catalog.parent-namespace=test_namespace
```

Fixes #22609

## Release notes

```markdown
# Iceberg
* Add support for reading from Unity catalog with REST catalog. ({issue}`22609`)
```

### CODE REVIEW

**Comment by decgowran** [comment]
@ebyhr do you know if this WIP or planning to be merged in the near future ?

File: docs/src/main/sphinx/object-storage/metastores.md [hunk 1]
```diff
@@ -497,6 +500,18 @@ iceberg.catalog.type=rest
 iceberg.rest-catalog.uri=http://iceberg-with-rest:8181
 ```
 
+`iceberg.security` must be `read_only` when connecting to Databricks Unity catalog
+using an Iceberg REST catalog:
```
Comment (anusudarsan) [inline docs/src/main/sphinx/object-storage/metastores.md:504]: maybe also include iceberg.rest-catalog.namespace in the example, even though its 
optional, with a non-default other than `main` ?

**Comment by MichaelKoch** [comment]
any roadmap to merge that PR ?

**Comment by decgowran** [comment]
> any roadmap to merge that PR ?

@MichaelKoch Im looking into alternative in k8s with https://github.com/duckdb/uc_catalog

**Comment by MichaelKoch** [comment]
@decgowran : thanks for the link .. 
i would love to see that this PR get merged to use trino + uc 
IMHO : a great partnership

File: docs/src/main/sphinx/object-storage/metastores.md [hunk 2]
```diff
@@ -470,6 +470,9 @@ following properties:
 * - `iceberg.rest-catalog.warehouse`
   - Warehouse identifier/location for the catalog (optional). Example:
     `s3://my_bucket/warehouse_location`
+* - `iceberg.rest-catalog.namespace`
```
Comment (ebyhr) [inline docs/src/main/sphinx/object-storage/metastores.md:473]: This config property will be used as "parent" parameter of REST call. `iceberg.rest-catalog.parent-namespace` might be better.

**Comment by ebyhr** [comment]
Please ignore the 1st commit and `ci / test (plugin/trino-iceberg)` failure. I'm going to handle in a separate PR.

File: docs/src/main/sphinx/object-storage/metastores.md (hunk 1 above)
Comment (ebyhr) [reply docs/src/main/sphinx/object-storage/metastores.md:504]: Actually, `main` isn't a default value. Added `iceberg.rest-catalog.namespace` config property to the example anyway.

**Review by raunaqmorarka — APPROVED** [review]

**Comment by decgowran** [comment]
Hi @ebyhr  Will this be available in the next Trino Release to test ?

File: docs/src/main/sphinx/object-storage/metastores.md [hunk 3]
```diff
@@ -497,6 +500,19 @@ iceberg.catalog.type=rest
 iceberg.rest-catalog.uri=http://iceberg-with-rest:8181
 ```
 
+`iceberg.security` must be `read_only` when connecting to Databricks Unity catalog
```
Comment (findinpath) [inline docs/src/main/sphinx/object-storage/metastores.md:503]: What happens when a user forgets ot set `iceberg.security=read_only` ?

Comment (ebyhr) [reply docs/src/main/sphinx/object-storage/metastores.md:503]: Write operations will result in the following error: 
```
Query 20241015_222748_00008_srrgj failed: Unable to process: Status: 405
Description: Method Not Allowed
```

**Pull request merged**